core/src/main/java/org/apache/iceberg/BaseTableScan.java

```diff
@@ -116,9 +133,20 @@ public CloseableIterable<FileScanTask> planFiles() {
           ExpressionUtil.toSanitizedString(filter()));
 
       Listeners.notifyAll(new ScanEvent(table().name(), snapshot.snapshotId(), filter(), schema()));
-
-      return doPlanFiles();
-
+      scanReporter().startScan();
+      Timer.Sample sample = scanReporter().scanMetrics().totalScanDuration().start();
```

Comment:
Tying the lifecycle of scan metrics to the scan reporter means that all of the scans created through this chain will have the same metrics collector, which is a correctness problem.

In the Java library, the `TableScan` interface uses a refinement pattern. Each `TableScan` is independent and you get a new table scan that has additional options when you call the refinement methods, like `select` or `filter`. For example,

```java
TableScan fullTableScan = table.newScan();
TableScan yesterday = fullTableScan.filter(Expressions.equal("date_col", yesterday()));
TableScan loginsYesterday = yesterday.filter(Expressions.equal("event_type", "login"));

CloseableIterable<FileScanTask> yesterdayTasks = yesterday.planFiles();
CloseableIterable<FileScanTask> loginTasks = loginsYesterday.planFiles();
```

Two scans are planned, but the metrics for both would use the same `ScanMetrics`.

Above, I suggested separating `ScanMetrics` from the reporter, which would solve this problem. Not all scans are planned, so you wouldn't want to eagerly create metrics, but the ones that are used should have a new one:

```java
  protected ScanMetrics metrics() {
    if (scanMetrics == null) {
      this.scanMetrics = scanReporter.newScanMetrics();
    }

    return scanMetrics;
  }

  public CloseableIterable<FileScanTask> planFiles(...) {
    ScanMetrics metrics = metrics();
    ...
  }
```

Reply:
thanks for bringing this up. I haven't thought about that. I'll rewrite the implementation and add some tests to make sure that each independent table scan produce their own metrics/report

Reply:
I double-checked and the old implementation was behaving correctly, because a new `ScanMetrics` instance was created when `ScanReporter#startScan()` was called.
However, I updated the `ScanReporter` implementation to how you suggested, as that is cleaner